### ISSUE TITLE
Error in test phase of mmod training

### ISSUE THREAD

**Issue by Benuraj** [davisking/dlib#341]
Hi,
I am training face imagenet data (with mmod loss) and training was successful. But in test phase I got following exception

done training

Error detected at line 745.
~/dlib/dnn/cuda_dlib.cu.
Error detected in function void dlib::cuda::affine_transform(dlib::tensor&, const dlib::tensor&, const dlib::tensor&, const dlib::tensor&).

Failing expression was A.nr()==B.nr() && B.nr()==src.nr() && A.nc()==B.nc() && B.nc()==src.nc() && A.k() ==B.k() && B.k()==src.k().

A.nr(): 556
B.nr(): 556
src.nr(): 1309
A.nc(): 148
B.nc(): 148
src.nc(): 478
A.k(): 16
B.k(): 16
src.k(): 16

Did anybody else faced similar issue? Is this something to do with layer configuration? I had changed config of dnn_mmod_ex to that of dnn_mmod_face_detection_ex.

**Comment by AndreaPisoni** [comment]
can you write down the network layers you are using in training and the ones you are using in detection?

**Comment by Benuraj** [comment]
the model in dnn_mmod_ex i modified to

`template <long num_filters, typename SUBNET> using con5d = con<num_filters,5,5,2,2,SUBNET>;
template <long num_filters, typename SUBNET> using con5  = con<num_filters,5,5,1,1,SUBNET>;

template <typename SUBNET> using downsampler  = relu<affine<con5d<32, relu<affine<con5d<32, relu<affine<con5d<16,SUBNET>>>>>>>>>;
template <typename SUBNET> using rcon5  = relu<affine<con5<45,SUBNET>>>;

using net_type = loss_mmod<con<1,9,9,1,1,rcon5<rcon5<rcon5<downsampler<input_rgb_image_pyramid<pyramid_down<6>>>>>>>>;`

after serializing the net i.e,
`serialize("mmod_network.dat") << net;`

I call test function as in example
`test_object_detection_function(net, images_test, face_boxes_test) << endl;`

**Comment by AndreaPisoni** [comment]
OK your problem is that you took the network for detection and you are using it for training. 

You can fix your issue by changing the affine layers to batch norm layers in TRAINING, and leave them as affine in testing, but I would suggest you read the docs on the layers as there are a few of these cases (dropout need to become mul in test etc.)

**Comment by Benuraj** [comment]
oh thanks a lot.

**Comment by davisking** [comment]
I should point out that you can use affine layers for training.  It's a weird thing to do and usually you wouldn't want to do it.  However, you can.  But by default they are in fully connected mode rather than convolutional mode, which is why you get this error since fully connected mode doesn't make sense for this network architecture.  

The reason it works in the other examples is because, when the affine layers are initialized by batch normalization layers, they get initialized to convolutional mode as well.

Anyway, don't do this.  It isn't what you want in any case.
